### PR TITLE
Load task from server if not found locally

### PR DESCRIPTION
**Pull request by raimund-schluessler**
This implements #606. When a requested task is not found locally, we try to load it from the server.

Todo:
- [x] We should also try to load the ancestor tasks (if any) of the requested task, so we can properly show the subtasks tree
- [x] If we are in a collection view (e.g. /nextcloud/index.php/apps/tasks/#/collections/all/tasks/Nextcloud-lrr0rvhdcoqopte5k43n6.ics), we don't know which calendar a task belongs to, so we have to query all calendars I guess. Or is there a better way to find a task only knowing the last part of its `uri` @georgehrke @skjnldsv ?
- [x] Don't loose subtask relations for tasks alread present in store when loading tasks by Uri or UID

What I don't like so much is that I have to watch the calendar (https://github.com/nextcloud/tasks/pull/608/files#diff-128fe9ee0ed5aeabe3c64281c67285c0R584), but it was the easiest way to load a task only after the calendars are loaded.

### CODE REVIEW

File: src/components/TheDetails.vue [hunk 1]
```diff
@@ -630,17 +631,30 @@ export default {
...
-						this.loading = false
-						throw error
-					})
+		async loadTask() {
+			if (this.task === undefined || this.task === null) {
+				const calendars = this.calendar ? [this.calendar] : this.calendars
+				for (const calendar of calendars) {
+					this.loading = true
+					const found = await this.getTaskByUri({ calendar: calendar, taskUri: this.$route.params.taskId })
+						.then((response) => {
```
[the remark this answers is not in the document]
Comment (raimund-schluessler) [reply src/components/TheDetails.vue]: So, like this 5615faf58c8ddfc7a3cdb12c045f6dcba23eca8c?

**Comment by raimund-schluessler** [comment]
This should be good to go now.

File: src/components/TheDetails.vue [hunk 2]
```diff
@@ -615,8 +628,36 @@ export default {
...
 		]),
 
+		async loadTask() {
+			if (this.task === undefined || this.task === null) {
+				const calendars = this.calendar ? [this.calendar] : this.calendars
+				for (const calendar of calendars) {
+					this.loading = true
+					try {
+						const found = await this.getTaskByUri({ calendar: calendar, taskUri: this.$route.params.taskId })
+							.then((response) => {
```
Comment (skjnldsv) [inline src/components/TheDetails.vue]: please don't mix try..;catch and .then/catch :)

File: src/store/tasks.js [hunk 3]
```diff
@@ -720,6 +721,96 @@ const actions = {
...
+	 * and commits the result
+	 *
+	 * @param {Object} context The store mutations
+	 * @param {Object} data Destructuring object
+	 * @param {Calendar} data.calendar The calendar
+	 * @param {String} data.taskUri The uri of the requested task
+	 * @returns {Promise}
+	 */
+	async getTaskByUri(context, { calendar, taskUri }) {
+		return calendar.dav.find(taskUri)
```
Comment (skjnldsv) [inline src/store/tasks.js]: try...catch? :)

File: src/components/TheDetails.vue (hunk 2 above)
Comment (raimund-schluessler) [reply src/components/TheDetails.vue]: What would be the correct way? Could you elaborate a bit, I am not really familiar with this.

Comment (skjnldsv) [reply src/components/TheDetails.vue]: It's usually good practice to not use .then/.catch anymore as you end up with chaining lots of promises :)
Using async/await is recommended ;)

Comment (raimund-schluessler) [reply src/components/TheDetails.vue]: Ok, I will have a look. Thanks!

Comment (raimund-schluessler) [reply src/components/TheDetails.vue]: Thanks for making me aware @skjnldsv. Looks much cleaner with async/await only 👍

**Comment by raimund-schluessler** [comment]
> See comments :)

Fixed it. Looks much better now with async/await.

File: src/store/tasks.js [hunk 4]
```diff
@@ -720,6 +721,88 @@ const actions = {
...
 
+	/**
+	 * Retrieves the task with the given uri from the given calendar
+	 * and commits the result
+	 *
+	 * @param {Object} context The store mutations
+	 * @param {Object} data Destructuring object
+	 * @param {Calendar} data.calendar The calendar
+	 * @param {String} data.taskUri The uri of the requested task
+	 * @returns {Promise}
```
Comment (skjnldsv) [inline src/store/tasks.js]:
```suggestion
	 * @returns {Task}
```

File: src/store/tasks.js [hunk 5]
```diff
@@ -720,6 +721,88 @@ const actions = {
...
+
+	/**
+	 * Retrieves the task with the given uid from the given calendar
+	 * and commits the result
+	 *
+	 * @param {Object} context The store mutations
+	 * @param {Object} data Destructuring object
+	 * @param {Calendar} data.calendar The calendar
+	 * @param {String} data.taskUid The uid of the requested task
+	 * @returns {Promise}
```
Comment (skjnldsv) [inline src/store/tasks.js]:
```suggestion
	 * @returns {Task}
```

File: src/store/tasks.js [hunk 6]
```diff
@@ -720,6 +721,88 @@ const actions = {
...
+		if (response) {
+			const task = new Task(response.data, calendar)
+			Vue.set(task, 'dav', response)
+			if (task.related) {
+				let parent = context.getters.getTaskByUid(task.related)
+				// If the parent is not found locally, we try to get it from the server.
+				if (!parent) {
+					parent = await context.dispatch('getTaskByUid', { calendar: calendar, taskUid: task.related })
+				}
+				context.commit('addTaskToParent', { task: task, parent: parent })
```
Comment (skjnldsv) [inline src/store/tasks.js]:
```suggestion
				context.commit('addTaskToParent', { task, parent })
```

File: src/store/tasks.js [hunk 7]
```diff
@@ -720,6 +721,88 @@ const actions = {
...
+	async getTaskByUri(context, { calendar, taskUri }) {
+		const response = await calendar.dav.find(taskUri)
+		if (response) {
+			const task = new Task(response.data, calendar)
+			Vue.set(task, 'dav', response)
+			if (task.related) {
+				let parent = context.getters.getTaskByUid(task.related)
+				// If the parent is not found locally, we try to get it from the server.
+				if (!parent) {
+					parent = await context.dispatch('getTaskByUid', { calendar: calendar, taskUid: task.related })
```
Comment (skjnldsv) [inline src/store/tasks.js]:
```suggestion
					parent = await context.dispatch('getTaskByUid', { calendar, taskUid: task.related })
```

File: src/store/tasks.js [hunk 8]
```diff
@@ -720,6 +721,88 @@ const actions = {
...
+
+			// In case we already have subtasks of this task in the store, add them as well.
+			const subTasksInStore = context.getters.getTasksByParent(task)
+			subTasksInStore.forEach(
+				subTask => {
+					context.commit('addTaskToParent', { task: subTask, parent: task })
+				}
+			)
+
+			context.commit('appendTasksToCalendar', { calendar: calendar, tasks: [task] })
```
Comment (skjnldsv) [inline src/store/tasks.js]:
```suggestion
			context.commit('appendTasksToCalendar', { calendar, tasks: [task] })
```

File: src/store/tasks.js [hunk 9]
```diff
@@ -720,6 +721,88 @@ const actions = {
...
+		const response = await findVTODObyUid(calendar, taskUid)
+		// We expect to only get zero or one task when we query by UID.
+		if (response.length) {
+			const task = new Task(response[0].data, calendar)
+			Vue.set(task, 'dav', response[0])
+			if (task.related) {
+				let parent = context.getters.getTaskByUid(task.related)
+				// If the parent is not found locally, we try to get it from the server.
+				if (!parent) {
+					parent = await context.dispatch('getTaskByUid', { calendar: calendar, taskUid: task.related })
```
Comment (skjnldsv) [inline src/store/tasks.js]:
```suggestion
					parent = await context.dispatch('getTaskByUid', { calendar, taskUid: task.related })
```

File: src/store/tasks.js [hunk 10]
```diff
@@ -720,6 +721,88 @@ const actions = {
...
+				}
+				context.commit('addTaskToParent', { task: task, parent: parent })
```
Comment (skjnldsv) [inline src/store/tasks.js]:
```suggestion
				context.commit('addTaskToParent', { task, parent })
```

File: src/store/tasks.js [hunk 11]
```diff
@@ -720,6 +721,88 @@ const actions = {
...
+
+			// In case we already have subtasks of this task in the store, add them as well.
+			const subTasksInStore = context.getters.getTasksByParent(task)
+			subTasksInStore.forEach(
+				subTask => {
+					context.commit('addTaskToParent', { task: subTask, parent: task })
+				}
+			)
+
+			context.commit('appendTasksToCalendar', { calendar: calendar, tasks: [task] })
```
Comment (skjnldsv) [inline src/store/tasks.js]:
```suggestion
			context.commit('appendTasksToCalendar', { calendar, tasks: [task] })
```

File: src/components/TheDetails.vue [hunk 12]
```diff
@@ -615,8 +628,28 @@ export default {
...
+			'getTaskByUri',
 		]),
 
+		async loadTask() {
+			if (this.task === undefined || this.task === null) {
+				const calendars = this.calendar ? [this.calendar] : this.calendars
+				for (const calendar of calendars) {
+					this.loading = true
+					try {
+						const task = await this.getTaskByUri({ calendar: calendar, taskUri: this.$route.params.taskId })
```
Comment (skjnldsv) [inline src/components/TheDetails.vue]: Much cleaner ! :+1: 
```suggestion
						const task = await this.getTaskByUri({ calendar, taskUri: this.$route.params.taskId })
```

File: src/store/tasks.js (hunk 4 above)
Comment (raimund-schluessler) [reply src/store/tasks.js]: Although strictly async functions return Promises as well 😉

**Comment by raimund-schluessler** [comment]
Rebased.

File: src/store/tasks.js [hunk 13]
```diff
@@ -720,6 +721,88 @@ const actions = {
...
+	 * @param {Object} data Destructuring object
+	 * @param {Calendar} data.calendar The calendar
+	 * @param {String} data.taskUri The uri of the requested task
+	 * @returns {Task}
+	 */
+	async getTaskByUri(context, { calendar, taskUri }) {
+		const response = await calendar.dav.find(taskUri)
+		if (response) {
+			const task = new Task(response.data, calendar)
+			Vue.set(task, 'dav', response)
```
Comment (skjnldsv) [inline src/store/tasks.js]: Not a big deal, more a small tip, but you can also use `this.$set` :)

Comment (skjnldsv) [reply src/store/tasks.js]: Ah no, we're in a store! Sorry!! :grin: :wink:

**Pull request merged**